flow/activities/maintenance_activity.go

```diff
@@ -0,0 +1,298 @@
+		ctx, 30*time.Second, func() {
+			slog.Info("Waiting for mirror to pause",
+				"mirror", mirrorInfo.Name, "workflowId", mirrorInfo.WorkflowId, "currentStatus", mirrorStatus.String())
+		},
+	)()
+
+	for mirrorStatus != protos.FlowStatus_STATUS_PAUSED {
+		time.Sleep(10 * time.Second)
+
+		activity.RecordHeartbeat(ctx, fmt.Sprintf("Waiting for mirror %s to pause", mirrorInfo.Name))
+		mirrorStatus, err = a.getMirrorStatus(ctx, mirrorInfo)
+		if err != nil {
+			slog.Error("Error querying mirror status for maintenance of previously running mirror",
+				"mirror", mirrorInfo.Name, "workflowId", mirrorInfo.WorkflowId, "error", err)
+		}
+	}
+	return true, nil
+}
+
+func (a *MaintenanceActivity) CleanupBackupedFlows(ctx context.Context) error {
+	_, err := a.CatalogPool.Exec(ctx, `
+		update maintenance.maintenance_flows
+			set state = $1,
+			restored_at = now(),
+			to_version = $2
+		where state = $3
+	`, mirrorStateRestored, peerdbenv.PeerDBVersionShaShort(), mirrorStateBackup)
+	return err
+}
+
+func (a *MaintenanceActivity) GetBackedUpFlows(ctx context.Context) (MaintenanceMirrorsInfo, error) {
+	rows, err := a.CatalogPool.Query(ctx, `
+		select flow_id, flow_name, workflow_id, flow_created_at, is_cdc
+		from maintenance.maintenance_flows
+		where state = $1
+	`, mirrorStateBackup)
+	if err != nil {
+		return MaintenanceMirrorsInfo{}, err
+	}
+
+	maintenanceMirrorItems, err := pgx.CollectRows(rows, func(row pgx.CollectableRow) (MaintenanceMirrorInfoItem, error) {
+		var info MaintenanceMirrorInfoItem
+		err := row.Scan(&info.Id, &info.Name, &info.WorkflowId, &info.CreatedAt, &info.IsCDC)
+		return info, err
+	})
+	if err != nil {
+		return MaintenanceMirrorsInfo{}, err
+	}
+
+	return MaintenanceMirrorsInfo{
+		Mirrors: maintenanceMirrorItems,
+	}, nil
+}
+
+func (a *MaintenanceActivity) ResumeMirror(ctx context.Context, mirrorInfo MaintenanceMirrorInfoItem) error {
+	mirrorStatus, err := a.getMirrorStatus(ctx, mirrorInfo)
+	if err != nil {
+		return err
+	}
+
+	if mirrorStatus != protos.FlowStatus_STATUS_PAUSED {
+		slog.Error("Cannot resume mirror that is not paused",
+			"mirror", mirrorInfo.Name, "workflowId", mirrorInfo.WorkflowId, "status", mirrorStatus.String())
+		return nil
+	}
+
+	// There can also be "workflow already completed" errors, what should we do in that case?
+	err = model.FlowSignal.SignalClientWorkflow(ctx, a.TemporalClient, mirrorInfo.WorkflowId, "", model.NoopSignal)
+	if err != nil {
+		slog.Error("Error signaling mirror to resume for maintenance",
+			"mirror", mirrorInfo.Name, "workflowId", mirrorInfo.WorkflowId, "error", err)
+		return err
+	}
+	return nil
+}
+
+func (a *MaintenanceActivity) DisableMaintenanceMode(ctx context.Context) error {
+	slog.Info("Disabling maintenance mode")
+	return peerdbenv.UpdatePeerDBMaintenanceModeEnabled(ctx, a.CatalogPool, false)
+}
+
+func (a *MaintenanceActivity) BackgroundAlerter(ctx context.Context) error {
+	defer shared.Interval(ctx, 30*time.Second, func() {
```

Comment:
this can be another ticker & be in the below `select`. See MaintainPull

Reply:
`PeerDBMaintenanceModeWaitAlertSeconds` can be a bit large, that is why I had a separate timeout for the hearbeat